# Analyzing borrowers’ risk of defaulting

Your project is to prepare a report for a bank’s loan division. You’ll need to find out if a customer’s marital status and number of children has an impact on whether they will default on a loan. The bank already has some data on customers’ credit worthiness.

Your report will be considered when building the **credit score** of a potential customer. The **credit score** is used to evaluate the ability of a potential borrower to repay their loan.

The purposes and hypotheses of the project:
1. Find out whether a customer's marital status impacts the loan default risk. My hypotheses for this is that a customer's marital status does impact the loan default risk. 
2. Find out whether the number of children a customer has impacts the loan default risk. My hypotheses for this is that the number of children a customer does impact the loan default risk. 

## Open the data file and have a look at the general information. 

In [1]:
# Loading all the libraries
import pandas as pd
import numpy as np


# Load the data
credit_score_data = pd.read_csv('/datasets/credit_scoring_eng.csv')

## Task 1. Data exploration

**Description of the data**
- `children` - the number of children in the family
- `days_employed` - work experience in days
- `dob_years` - client's age in years
- `education` - client's education
- `education_id` - education identifier
- `family_status` - marital status
- `family_status_id` - marital status identifier
- `gender` - gender of the client
- `income_type` - type of employment
- `debt` - was there any debt on loan repayment
- `total_income` - monthly income
- `purpose` - the purpose of obtaining a loan

Now let's explore our data. First we want to see how many columns and rows it has.

In [2]:
# Let's see how many rows and columns our dataset has 
print(credit_score_data.shape)
# and the general information about the data set
credit_score_data.info()

(21525, 12)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       19351 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        19351 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


It shows that the data set has 21525 rows and 12 columns. The columns have different types such as int, float and object. Next, we will look at the first 10 rows to check for potential issues with the data.

In [3]:
# Let's look at the first 10 rows of the dataset
credit_score_data.head(10)

children  days_employed  dob_years            education  education_id  \
0         1   -8437.673028         42    bachelor's degree             0   
1         1   -4024.803754         36  secondary education             1   
2         0   -5623.422610         33  Secondary Education             1   
3         3   -4124.747207         32  secondary education             1   
4         0  340266.072047         53  secondary education             1   
5         0    -926.185831         27    bachelor's degree             0   
6         0   -2879.202052         43    bachelor's degree             0   
7         0    -152.779569         50  SECONDARY EDUCATION             1   
8         2   -6929.865299         35    BACHELOR'S DEGREE             0   
9         0   -2188.756445         41  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   
5  civil partnership                 1      M    business     0     40922.170   
6            married                 0      F    business     0     38484.156   
7            married                 0      M    employee     0     21731.829   
8  civil partnership                 1      F    employee     0     15337.093   
9            married                 0      M    employee     0     23108.150   

                               purpose  
0                purchase of the house  
1                         car purchase  
2                purchase of the house  
3              supplementary education  
4                    to have a wedding  
5                purchase of the house  
6                 housing transactions  
7                            education  
8                     having a wedding  
9  purchase of the house for my family

In [4]:
# let's print the first N rows
N = 15
credit_score_data.head(N)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42    bachelor's degree             0   
1          1   -4024.803754         36  secondary education             1   
2          0   -5623.422610         33  Secondary Education             1   
3          3   -4124.747207         32  secondary education             1   
4          0  340266.072047         53  secondary education             1   
5          0    -926.185831         27    bachelor's degree             0   
6          0   -2879.202052         43    bachelor's degree             0   
7          0    -152.779569         50  SECONDARY EDUCATION             1   
8          2   -6929.865299         35    BACHELOR'S DEGREE             0   
9          0   -2188.756445         41  secondary education             1   
10         2   -4171.483647         36    bachelor's degree             0   
11         0    -792.701887         40  secondary education             1   
12         0            NaN         65  secondary education             1   
13         0   -1846.641941         54         some college             2   
14         0   -1844.956182         56    bachelor's degree             0   

        family_status  family_status_id gender income_type  debt  \
0             married                 0      F    employee     0   
1             married                 0      F    employee     0   
2             married                 0      M    employee     0   
3             married                 0      M    employee     0   
4   civil partnership                 1      F     retiree     0   
5   civil partnership                 1      M    business     0   
6             married                 0      F    business     0   
7             married                 0      M    employee     0   
8   civil partnership                 1      F    employee     0   
9             married                 0      M    employee     0   
10            married                 0      M    business     0   
11            married                 0      F    employee     0   
12  civil partnership                 1      M     retiree     0   
13            married                 0      F    employee     0   
14  civil partnership                 1      F    business     1   

    total_income                              purpose  
0      40620.102                purchase of the house  
1      17932.802                         car purchase  
2      23341.752                purchase of the house  
3      42820.568              supplementary education  
4      25378.572                    to have a wedding  
5      40922.170                purchase of the house  
6      38484.156                 housing transactions  
7      21731.829                            education  
8      15337.093                     having a wedding  
9      23108.150  purchase of the house for my family  
10     18230.959                      buy real estate  
11     12331.077           buy commercial real estate  
12           NaN                    to have a wedding  
13     20873.317                         car purchase  
14     26420.466          buy residential real estate

Looking at the first 15 rows of the dataset, we see that there are no obvious issues with column named `children`, `dob_years`, `education`, `education_id`, `family_status`, `family_status_id`, `gender`, `income_type` and `debt`.

But the column `days_employed` has many negative values, which does not make sense since this column represents work experience in days. It also has some missing values. The column `education` has inconsistent lower/upper case characters for the same category. The column `total_income` has some missing values. The column `purpose` has different wording for the same category.     

In [5]:
# Get info on data
print(credit_score_data.isna().sum())

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64


The column named `days_employed` has 2174 missing values and the column named `total_income` also has 2174 missing values. All other columns do not have any missing values.  

In [6]:
# Let's look at the filtered table with missing values in the the first column with missing data

print(credit_score_data[credit_score_data['days_employed'].isna()]['days_employed'])

print(credit_score_data[credit_score_data['total_income'].isna()]['total_income'])

12      NaN
26      NaN
29      NaN
41      NaN
55      NaN
         ..
21489   NaN
21495   NaN
21497   NaN
21502   NaN
21510   NaN
Name: days_employed, Length: 2174, dtype: float64
12      NaN
26      NaN
29      NaN
41      NaN
55      NaN
         ..
21489   NaN
21495   NaN
21497   NaN
21502   NaN
21510   NaN
Name: total_income, Length: 2174, dtype: float64


By looking at the row indices of the first and last five rows of missing values from the columns `days_employed` and `total_income`, we suspect that missing values seem symmetric, which means if a row is missing`days_employed`, it also has `total_income` missing. But we are not sure about this assumption yet. We need to conduct further investigations to find out whether it is true. Specifically, we need to find out whether missing values from `days_employed` and `total_income` come from the same rows. 

In [7]:
# Let's apply multiple conditions for filtering data and look at the number of rows in the filtered table.
print(credit_score_data[credit_score_data['days_employed'].isna()]['total_income'].isna().sum())

print(credit_score_data[credit_score_data['total_income'].isna()]['days_employed'].isna().sum())

2174
2174


In [8]:
perct_missing = credit_score_data['days_employed'].isna().sum()/credit_score_data.shape[0]
print(f'The percentage of missing values in the data set is: {perct_missing:.2%}')

The percentage of missing values in the data set is: 10.10%


**Intermediate conclusion**

1. The above results shows that missing values from `days_employed` and `total_income` are symmetric because when `days_employed` has missing values, `total_income` also have exactly the same number of missing values, and vice versa. 

2. The data set has about 10.10% missing values. Because the data set is quite large and the percentage of missing values is relatively high, I think it is better not to drop all rows with missing values. We should consider whether the missing data could be due to the specific client characteristic, such as a customer's marital status, the number of children a customer has and whether the customer has ever defaulted on a loan. Specifically, we should check whether there's any dependence missing values have on the value of other indicators with the columns with identified specific client characteristic.     

In [9]:
# Let's investigate clients who do not have data on identified characteristic and the column with the missing values
rows_with_missing_values = credit_score_data[credit_score_data['days_employed'].isna()].reset_index(drop=True)
# print(rows_with_missing_values)
columns_with_missing_values = credit_score_data[credit_score_data['days_employed'].isna()][['days_employed', 'total_income']].reset_index(drop=True)
print(columns_with_missing_values)

      days_employed  total_income
0               NaN           NaN
1               NaN           NaN
2               NaN           NaN
3               NaN           NaN
4               NaN           NaN
...             ...           ...
2169            NaN           NaN
2170            NaN           NaN
2171            NaN           NaN
2172            NaN           NaN
2173            NaN           NaN

[2174 rows x 2 columns]


In [10]:
size_dataset_debt = credit_score_data.groupby('debt').size()
default_dataset = size_dataset_debt[1]/(size_dataset_debt[0]+size_dataset_debt[1])
print(f'The loan default rate in the entire data set is {default_dataset:.2%}')

size_missing_data_debt = rows_with_missing_values.groupby('debt').size()
default_missing_data = size_missing_data_debt[1]/(size_missing_data_debt[0]+size_missing_data_debt[1])
print(f'The loan default rate in the data set with missing values is {default_missing_data:.2%}')

# https://pandas.pydata.org/docs/reference/api/pandas.pivot_table.html

The loan default rate in the entire data set is 8.09%
The loan default rate in the data set with missing values is 7.82%


The loan default rate in the entire data set is not very different from the loan default rate in the data set with missing values. Thus, the missing values do not seem to have any dependence on whether the loan has defaulted or not.

In [11]:
print("The percentage of each category in the entire data set based on grouping the 'family_status' column:")
credit_score_data.groupby('family_status').size().transform(lambda x: x/x.sum())

The percentage of each category in the entire data set based on grouping the 'family_status' column:


family_status
civil partnership    0.194053
divorced             0.055517
married              0.575145
unmarried            0.130685
widow / widower      0.044599
dtype: float64

In [12]:
print("The percentage of each category in the sub-dataset with missing values based on grouping the 'family_status' column:")
rows_with_missing_values.groupby('family_status').size().transform(lambda x: x/x.sum())

The percentage of each category in the sub-dataset with missing values based on grouping the 'family_status' column:


family_status
civil partnership    0.203312
divorced             0.051518
married              0.568997
unmarried            0.132475
widow / widower      0.043698
dtype: float64

The distribution of the data set with missing values does not seem to have any dependence on marital status of customers.

In [13]:
print("The percentage of each category based on grouping the 'children' column in the entire data set:")
credit_score_data.groupby('children').size().transform(lambda x: x/x.sum())

The percentage of each category based on grouping the 'children' column in the entire data set:


children
-1     0.002184
 0     0.657329
 1     0.223833
 2     0.095470
 3     0.015331
 4     0.001905
 5     0.000418
 20    0.003531
dtype: float64

In [14]:
print("The percentage of each category based on grouping the 'children' column in the sub-dataset with missing values:")
rows_with_missing_values.groupby('children').size().transform(lambda x: x/x.sum())

The percentage of each category based on grouping the 'children' column in the sub-dataset with missing values:


children
-1     0.001380
 0     0.661914
 1     0.218491
 2     0.093836
 3     0.016559
 4     0.003220
 5     0.000460
 20    0.004140
dtype: float64

The above calculation shows that the distribution of the data set with missing values are quite similar with the distribution of the entire data set in terms of the number of children in the family, except that when the number of children is 4, the percentage from the missing values sub-dataset is much higher than that from the entire data set. However, because the percentage of clients having 4 children is quite low, we can conclude that the missing values do not depend on the `children` column.  

In [15]:
print("The percentage of each category based on grouping the 'income_type' column in the entire data set:")
credit_score_data.groupby('income_type').size().transform(lambda x: x/x.sum())

The percentage of each category based on grouping the 'income_type' column in the entire data set:


income_type
business                       0.236237
civil servant                  0.067782
employee                       0.516562
entrepreneur                   0.000093
paternity / maternity leave    0.000046
retiree                        0.179141
student                        0.000046
unemployed                     0.000093
dtype: float64

In [16]:
print("The percentage of each category based on grouping the 'income_type' column in the sub-dataset with missing values:")
rows_with_missing_values.groupby('income_type').size().transform(lambda x: x/x.sum())

The percentage of each category based on grouping the 'income_type' column in the sub-dataset with missing values:


income_type
business         0.233671
civil servant    0.067617
employee         0.508280
entrepreneur     0.000460
retiree          0.189972
dtype: float64

The above calculation shows that the distribution of the data set with missing values are quite different from the distribution of the entire data set in terms of the customers' income types. The subset of the data with missing values does not have anyone with income types of paternity/maternity leave, student, or unemployed. It is very likely that missing values are related to income type.  

**Possible reasons for missing values in data and intermediate conclusion**

It is hard to determine exactly why some values might be missing. Based on the above analysis, it is possible that missing values are not random and they might be related to income type or age of the clients.  

Let's start checking whether the missing values are random.

## Data transformation

In [17]:
# Let's see all values in education column to check if and what spellings will need to be fixed
print(credit_score_data['education'].unique())

["bachelor's degree" 'secondary education' 'Secondary Education'
 'SECONDARY EDUCATION' "BACHELOR'S DEGREE" 'some college'
 'primary education' "Bachelor's Degree" 'SOME COLLEGE' 'Some College'
 'PRIMARY EDUCATION' 'Primary Education' 'Graduate Degree'
 'GRADUATE DEGREE' 'graduate degree']


In [18]:
# Fix the registers if required
credit_score_data['education'] = credit_score_data['education'].str.lower()

In [19]:
# Checking all the values in the column to make sure we fixed them

print(credit_score_data['education'].str.lower().unique())

["bachelor's degree" 'secondary education' 'some college'
 'primary education' 'graduate degree']


In [20]:
# Let's see the distribution of values in the `children` column
print(credit_score_data['children'].unique())

[ 1  0  3  2 -1  4 20  5]


In [21]:
print(sum(credit_score_data['children']== -1))

fraction_children_negative = sum(credit_score_data['children']== -1)/(credit_score_data.shape[0])

print(f'The percentage of rows with negative values in the children column is: {fraction_children_negative:.2%}')

47
The percentage of rows with negative values in the children column is: 0.22%


There are 47 rows with negative values (-1) in the `children` column, which is about 0.22% of the dataset. It is hard to tell why those values are negative. Because the percentage of rows with negative values in the `children` column is very small, I think it is safe to drop those rows. 

In [22]:
# [fix the data based on your decision]
credit_score_data = credit_score_data.loc[credit_score_data["children"] != -1]

In [23]:
# Checking the `children` column again to make sure it's all fixed
print(credit_score_data['children'].unique())

# credit_score_data.shape
# print (21525-credit_score_data.shape[0])
# reference: https://www.shanelynn.ie/pandas-drop-delete-dataframe-rows-columns/

[ 1  0  3  2  4 20  5]


In [24]:
# Find problematic data in `days_employed`, if they exist, and calculate the percentage

print(credit_score_data['days_employed'].unique())

print(sum(credit_score_data['days_employed']<0))

perct_negative_days_employed = sum(credit_score_data['days_employed']<0)/credit_score_data.shape[0]

print(f'The percentage of days employed with negative values is: {perct_negative_days_employed:.2%}')

[-8437.67302776 -4024.80375385 -5623.42261023 ... -2113.3468877
 -3112.4817052  -1984.50758853]
15869
The percentage of days employed with negative values is: 73.88%


We see that data in the `days_employed` column has many negative values. The percentage of days employed with negative values is as high as 73.88%. This might be caused by some technical issues which added a negative sign to the values. I think the most reasonable way to correct the data in the `days_employed` column is to change the negative numbers to positive numbers with the same absolute values.    

In [25]:
# Address the problematic values, if they exist
credit_score_data['days_employed'] = credit_score_data['days_employed'].abs()

In [26]:
# Check the result - make sure it's fixed
print(credit_score_data['days_employed'].unique())

[8437.67302776 4024.80375385 5623.42261023 ... 2113.3468877  3112.4817052
 1984.50758853]


Let's now look at the client's age and whether there are any issues there.

In [27]:
# Check the `dob_years` for suspicious values and count the percentage
print(credit_score_data['dob_years'].unique())
print(credit_score_data['dob_years'].min())
print(credit_score_data['dob_years'].max())


[42 36 33 32 53 27 43 50 35 41 40 65 54 56 26 48 24 21 57 67 28 63 62 47
 34 68 25 31 30 20 49 37 45 61 64 44 52 46 23 38 39 51  0 59 29 60 55 58
 71 22 73 66 69 19 72 70 74 75]
0
75


The age ranges from 0 to 75. But 0 cannot be someone's age. Now let's check how many rows with age = 0.

In [28]:
print(sum(credit_score_data['dob_years'] == 0))

101


There are 101 rows with clients' age that is 0. I suspect that age = 0 means age is missing. Because there are not that many rows with age = 0, I think it is safe to remove those rows with missing values.  

In [29]:
# Address the issues in the `dob_years` column, if they exist
credit_score_data = credit_score_data.loc[credit_score_data["dob_years"] > 0]

In [30]:
# Check the result - make sure it's fixed
print(credit_score_data['dob_years'].min())

19


Now it shows that the minimum age of all clients is 19, which is reasonable. We can move on to check the `family_status` column.  

In [31]:
# Let's see the values for the column
print(credit_score_data['family_status'].unique())

['married' 'civil partnership' 'widow / widower' 'divorced' 'unmarried']


I do not see any issues with the `family_status` column. Now let's move on to check the `gender` column.

In [32]:
# Let's see the values in the column
print(credit_score_data['gender'].unique())

['F' 'M' 'XNA']


There are three unique values in the `gender` column: F, M and XNA. I think XNA means missing values. Let's find out how many missing values the `gender` column has.   

In [33]:
# Address the problematic values, if they exist
print(sum(credit_score_data['gender'] == 'XNA'))

1


There is only one row that has gender = XNA. We can remove this row from the dataset.

In [34]:
credit_score_data = credit_score_data.loc[credit_score_data["gender"] != 'XNA']

In [35]:
# Check the result - make sure it's fixed
print(credit_score_data['gender'].unique())

['F' 'M']


Now the `gender` column has two unique values: F, M. This looks good! Next, let's check the `income_type` column.

In [36]:
# Let's see the values in the column
print(credit_score_data['income_type'].unique())

['employee' 'retiree' 'business' 'civil servant' 'unemployed'
 'entrepreneur' 'student' 'paternity / maternity leave']


I do not see any issues with the `income_type` column. Now let's see if we have any duplicates in our data.

In [37]:
# Checking duplicates
credit_score_data.duplicated().sum()

71

In [38]:
# Address the duplicates, if they exist
credit_score_data = credit_score_data.drop_duplicates().reset_index(drop=True)

In [39]:
# Last check whether we have any duplicates
credit_score_data.duplicated().sum()

0

The data set has 71 duplicated rows and they are dropped from the data set.  

In [40]:
# Check the size of the dataset that you now have after your first manipulations with it
credit_score_data.shape

(21305, 12)

In [41]:
perct_rows_removed = (21525-21305)/21525
print(f'The percentage of rows removed from the dataset is: {perct_rows_removed:.2%}')

The percentage of rows removed from the dataset is: 1.02%


In [42]:
credit_score_data.head(15)

children  days_employed  dob_years            education  education_id  \
0          1    8437.673028         42    bachelor's degree             0   
1          1    4024.803754         36  secondary education             1   
2          0    5623.422610         33  secondary education             1   
3          3    4124.747207         32  secondary education             1   
4          0  340266.072047         53  secondary education             1   
5          0     926.185831         27    bachelor's degree             0   
6          0    2879.202052         43    bachelor's degree             0   
7          0     152.779569         50  secondary education             1   
8          2    6929.865299         35    bachelor's degree             0   
9          0    2188.756445         41  secondary education             1   
10         2    4171.483647         36    bachelor's degree             0   
11         0     792.701887         40  secondary education             1   
12         0            NaN         65  secondary education             1   
13         0    1846.641941         54         some college             2   
14         0    1844.956182         56    bachelor's degree             0   

        family_status  family_status_id gender income_type  debt  \
0             married                 0      F    employee     0   
1             married                 0      F    employee     0   
2             married                 0      M    employee     0   
3             married                 0      M    employee     0   
4   civil partnership                 1      F     retiree     0   
5   civil partnership                 1      M    business     0   
6             married                 0      F    business     0   
7             married                 0      M    employee     0   
8   civil partnership                 1      F    employee     0   
9             married                 0      M    employee     0   
10            married                 0      M    business     0   
11            married                 0      F    employee     0   
12  civil partnership                 1      M     retiree     0   
13            married                 0      F    employee     0   
14  civil partnership                 1      F    business     1   

    total_income                              purpose  
0      40620.102                purchase of the house  
1      17932.802                         car purchase  
2      23341.752                purchase of the house  
3      42820.568              supplementary education  
4      25378.572                    to have a wedding  
5      40922.170                purchase of the house  
6      38484.156                 housing transactions  
7      21731.829                            education  
8      15337.093                     having a wedding  
9      23108.150  purchase of the house for my family  
10     18230.959                      buy real estate  
11     12331.077           buy commercial real estate  
12           NaN                    to have a wedding  
13     20873.317                         car purchase  
14     26420.466          buy residential real estate

After checking each column, processing the abnormal values, and dropping duplicated rows, the new dataset now has 21305 rows and 12 columns. The percentage of rows removed from the dataset is about 1.02%. 

# Working with missing values

To speed up working with some data, I want to work with dictionaries for some values, where IDs are provided. It is obvious that values from the columns `education_id` and `education` are key value pairs and they can be used to create a dictionary. Moreover, the columns `family_status_id` and `family_status` can be used to create another dictionary. 

The dictionaries I will work with are shown below:

In [43]:
education_id_education = credit_score_data.groupby(['education_id', 'education']).size()
education_id_education

education_id  education          
0             bachelor's degree       5206
1             secondary education    15071
2             some college             740
3             primary education        282
4             graduate degree            6
dtype: int64

In [44]:
# Find the dictionaries
dict_education = {0:"bachelor's degree", 
                  1:"secondary education", 
                  2:"some college", 
                  3:"primary education", 
                  4:"graduate degree"}

In [45]:
family_status_id_family_status = credit_score_data.groupby(['family_status_id', 'family_status']).size()
family_status_id_family_status

family_status_id  family_status    
0                 married              12261
1                 civil partnership     4124
2                 widow / widower        950
3                 divorced              1181
4                 unmarried             2789
dtype: int64

In [46]:
# Find the dictionaries
dict_family_status = {0:"married", 
                  1:"civil partnership", 
                  2:"widow / widower", 
                  3:"divorced", 
                  4:"unmarried"}

### Restoring missing values in `total_income`

First, we want to find ways to impute the missing values in the `total_income` column. Total income might be associated with a person's days of employment, age, education level, gender and income type. With reasonable simplicity, we will: 1) create age categories for clients. 2) impute missing values in the `total_income` column using the median income for each age category. 

In [47]:
print(credit_score_data['dob_years'].value_counts().sort_index())

19     14
20     51
21    111
22    183
23    251
24    263
25    357
26    407
27    492
28    501
29    544
30    536
31    557
32    507
33    579
34    598
35    615
36    554
37    534
38    595
39    572
40    605
41    603
42    594
43    511
44    543
45    496
46    469
47    477
48    535
49    508
50    511
51    445
52    484
53    458
54    474
55    442
56    483
57    454
58    454
59    442
60    374
61    353
62    348
63    268
64    259
65    193
66    182
67    167
68     99
69     84
70     65
71     56
72     33
73      8
74      6
75      1
Name: dob_years, dtype: int64


In [48]:
# Let's write a function that calculates the age category

def age_group(age):
    
    """
    The function returns the age group according to the age values, using the following rules:
    'young adult' for age <= 30
    'adult'  for 30 < age <= 64
    'retired'  for age > 64
    """
    if age <= 30:
        return 'young adult'
    elif age <= 64:
        return 'adult'
    else:
        return 'retired'

In [49]:
# Test if the function works
age1 = 27
print(age_group(age1))

age2 = 37
print(age_group(age2))

age3 = 67
print(age_group(age3))

young adult
adult
retired


In [50]:
# Creating new column based on function

credit_score_data['age_group'] = credit_score_data['dob_years'].apply(age_group)
rows_with_missing_values['age_group'] =rows_with_missing_values['dob_years'].apply(age_group)

In [51]:
# Checking how values in the new column

credit_score_data['age_group']

0          adult
1          adult
2          adult
3          adult
4          adult
          ...   
21300      adult
21301    retired
21302      adult
21303      adult
21304      adult
Name: age_group, Length: 21305, dtype: object

In [52]:
rows_with_missing_values = credit_score_data[credit_score_data['days_employed'].isna()].reset_index(drop=True)
print(rows_with_missing_values)

      children  days_employed  dob_years            education  education_id  \
0            0            NaN         65  secondary education             1   
1            0            NaN         41  secondary education             1   
2            0            NaN         63  secondary education             1   
3            0            NaN         50  secondary education             1   
4            0            NaN         54  secondary education             1   
...        ...            ...        ...                  ...           ...   
2085         2            NaN         47  secondary education             1   
2086         1            NaN         50  secondary education             1   
2087         0            NaN         48    bachelor's degree             0   
2088         1            NaN         42  secondary education             1   
2089         2            NaN         28  secondary education             1   

          family_status  family_status_id gender   

In [53]:
# Create a table that only has data without missing values. 
# This data will be used to restore the missing values.
credit_score_data_no_missing = credit_score_data.dropna(axis='rows').reset_index(drop=True)

In [54]:
# Create a table without missing values and print a few of its rows to make sure it looks fine
print(credit_score_data_no_missing.head(15))

    children  days_employed  dob_years            education  education_id  \
0          1    8437.673028         42    bachelor's degree             0   
1          1    4024.803754         36  secondary education             1   
2          0    5623.422610         33  secondary education             1   
3          3    4124.747207         32  secondary education             1   
4          0  340266.072047         53  secondary education             1   
5          0     926.185831         27    bachelor's degree             0   
6          0    2879.202052         43    bachelor's degree             0   
7          0     152.779569         50  secondary education             1   
8          2    6929.865299         35    bachelor's degree             0   
9          0    2188.756445         41  secondary education             1   
10         2    4171.483647         36    bachelor's degree             0   
11         0     792.701887         40  secondary education             1   

In [55]:
# Look at the mean values for income based on your identified factors
credit_score_data_no_missing.pivot_table(index='age_group', 
                                        values='total_income', 
                                        aggfunc= 'mean')

total_income
age_group                
adult        27305.187614
retired      21546.199144
young adult  25810.971051

In [56]:
# Look at the median values for income based on your identified factors
credit_score_data_no_missing.pivot_table(index='age_group', 
                                        values='total_income', 
                                        aggfunc= 'median')

total_income
age_group                
adult           23547.067
retired         18434.205
young adult     22955.474

The distribution of income is right skewed because mean is larger than the median for each age group. Thus, median income represents the average population income better. It is better to use median income for imputation for the total income.  

In [57]:
#  Write a function that we will use for filling in missing values

def total_income_fill_na(row):
    age_group = row['age_group']
    if age_group == 'young adult':
        return 22955.474
    elif age_group == 'adult':
        return 23547.067
    else:
        return 18434.205
        

In [58]:
# print(rows_with_missing_values)

In [59]:
# Check if it works
# age_group1 = 'young adult'
row1 = rows_with_missing_values.loc[0]
# print(row1)
print(total_income_fill_na(row1))

row2 = rows_with_missing_values.loc[1]
# print(row1)
print(total_income_fill_na(row2))

row3 = rows_with_missing_values.loc[2089]
# print(row1)
print(total_income_fill_na(row3))

18434.205
23547.067
22955.474


In [60]:
# Apply it to every row

rows_with_missing_values['total_income'] = rows_with_missing_values.apply(total_income_fill_na, axis=1)

In [61]:
# Check if we got any errors
rows_with_missing_values

children  days_employed  dob_years            education  education_id  \
0            0            NaN         65  secondary education             1   
1            0            NaN         41  secondary education             1   
2            0            NaN         63  secondary education             1   
3            0            NaN         50  secondary education             1   
4            0            NaN         54  secondary education             1   
...        ...            ...        ...                  ...           ...   
2085         2            NaN         47  secondary education             1   
2086         1            NaN         50  secondary education             1   
2087         0            NaN         48    bachelor's degree             0   
2088         1            NaN         42  secondary education             1   
2089         2            NaN         28  secondary education             1   

          family_status  family_status_id gender    income_type  debt  \
0     civil partnership                 1      M        retiree     0   
1               married                 0      M  civil servant     0   
2             unmarried                 4      F        retiree     0   
3               married                 0      F  civil servant     0   
4     civil partnership                 1      F        retiree     1   
...                 ...               ...    ...            ...   ...   
2085            married                 0      M       business     0   
2086  civil partnership                 1      F       employee     0   
2087            married                 0      F       business     0   
2088            married                 0      F       employee     0   
2089            married                 0      F       employee     0   

      total_income                   purpose    age_group  
0        18434.205         to have a wedding      retired  
1        23547.067                 education        adult  
2        23547.067    building a real estate        adult  
3        23547.067  second-hand car purchase        adult  
4        23547.067         to have a wedding        adult  
...            ...                       ...          ...  
2085     23547.067         purchase of a car        adult  
2086     23547.067          wedding ceremony        adult  
2087     23547.067       building a property        adult  
2088     23547.067    building a real estate        adult  
2089     22955.474              car purchase  young adult  

[2090 rows x 13 columns]

In [62]:
# Checking the number of entries in the columns

len(rows_with_missing_values['total_income'])

2090

###  Restoring values in `days_employed`

If a person work for 40 years, then equavalently he/she works 14600 days. 

In [63]:
credit_score_data_no_missing[credit_score_data_no_missing['days_employed']>=14600]['days_employed']

# credit_score_data_no_missing['days_employed'].sort_values().tail(100)

4        340266.072047
17       400281.136913
23       338551.952911
24       363548.489348
27       335581.668515
             ...      
19196    338904.866406
19199    386497.714078
19200    362161.054124
19208    373995.710838
19211    343937.404131
Name: days_employed, Length: 3438, dtype: float64

Notice that in the dataset, there are 3438 clients who reported themselves working more than 300000 days, which is about 822 years! That is impossile! There must be some data entry errors for clients with unusually large values for days imployed. It is hard to tell why those values are unrealisticly large. So I decide to drop rows with `days_employed` >= 14600. 

In [64]:
credit_score_data_no_missing = credit_score_data_no_missing[credit_score_data_no_missing['days_employed'] <= 14600]

In [65]:
# Distribution of `days_employed` medians based on your identified parameters
credit_score_data_no_missing.pivot_table(index='age_group', 
                                        values='days_employed', 
                                        aggfunc= 'median')

days_employed
age_group                 
adult          1892.676984
retired        2761.484453
young adult    1043.334773

In [66]:
# Distribution of `days_employed` means based on your identified parameters
credit_score_data_no_missing.pivot_table(index='age_group', 
                                        values='days_employed', 
                                        aggfunc= 'mean')

days_employed
age_group                 
adult          2614.090656
retired        3817.982948
young adult    1279.049310

The distribution of the number of days employed is right skewed because mean is larger than the median for each age group. Thus, median income represents the average number of days employed better. It is better to use median for imputation for the `days_employed` column.  

In [67]:
# Let's write a function that calculates means or medians (depending on your decision) based on your identified parameter

def days_employed_fill_na(row):
    age_group = row['age_group']
    if age_group == 'young adult':
        return 1043.334773
    elif age_group == 'adult':
        return 1892.676984
    else:
        return 2761.484453

In [68]:
# Check that the function works

row1 = rows_with_missing_values.loc[0]
# print(row1)
print(days_employed_fill_na(row1))

row2 = rows_with_missing_values.loc[1]
# print(row1)
print(days_employed_fill_na(row2))

row3 = rows_with_missing_values.loc[2089]
# print(row1)
print(days_employed_fill_na(row3))

2761.484453
1892.676984
1043.334773


In [69]:
# Apply function to the days_employed

rows_with_missing_values['days_employed'] = rows_with_missing_values.apply(days_employed_fill_na, axis=1)

In [70]:
# Replacing missing values

rows_with_missing_values

children  days_employed  dob_years            education  education_id  \
0            0    2761.484453         65  secondary education             1   
1            0    1892.676984         41  secondary education             1   
2            0    1892.676984         63  secondary education             1   
3            0    1892.676984         50  secondary education             1   
4            0    1892.676984         54  secondary education             1   
...        ...            ...        ...                  ...           ...   
2085         2    1892.676984         47  secondary education             1   
2086         1    1892.676984         50  secondary education             1   
2087         0    1892.676984         48    bachelor's degree             0   
2088         1    1892.676984         42  secondary education             1   
2089         2    1043.334773         28  secondary education             1   

          family_status  family_status_id gender    income_type  debt  \
0     civil partnership                 1      M        retiree     0   
1               married                 0      M  civil servant     0   
2             unmarried                 4      F        retiree     0   
3               married                 0      F  civil servant     0   
4     civil partnership                 1      F        retiree     1   
...                 ...               ...    ...            ...   ...   
2085            married                 0      M       business     0   
2086  civil partnership                 1      F       employee     0   
2087            married                 0      F       business     0   
2088            married                 0      F       employee     0   
2089            married                 0      F       employee     0   

      total_income                   purpose    age_group  
0        18434.205         to have a wedding      retired  
1        23547.067                 education        adult  
2        23547.067    building a real estate        adult  
3        23547.067  second-hand car purchase        adult  
4        23547.067         to have a wedding        adult  
...            ...                       ...          ...  
2085     23547.067         purchase of a car        adult  
2086     23547.067          wedding ceremony        adult  
2087     23547.067       building a property        adult  
2088     23547.067    building a real estate        adult  
2089     22955.474              car purchase  young adult  

[2090 rows x 13 columns]

In [71]:
# Check the entries in all columns - make sure we fixed all missing values
credit_score_data = credit_score_data_no_missing.append(rows_with_missing_values).reset_index(drop=True)

print(credit_score_data.isna().sum())

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
age_group           0
dtype: int64


In [72]:
credit_score_data.info()
credit_score_data.shape

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17867 entries, 0 to 17866
Data columns (total 13 columns):
children            17867 non-null int64
days_employed       17867 non-null float64
dob_years           17867 non-null int64
education           17867 non-null object
education_id        17867 non-null int64
family_status       17867 non-null object
family_status_id    17867 non-null int64
gender              17867 non-null object
income_type         17867 non-null object
debt                17867 non-null int64
total_income        17867 non-null float64
purpose             17867 non-null object
age_group           17867 non-null object
dtypes: float64(2), int64(5), object(6)
memory usage: 1.8+ MB


(17867, 13)

After cleaning up the dataset, the processed dataset now has 17867 rows and 13 columns, without any missing values in all columns. 

## Categorization of data

First, let's look at the `family_status` column. 

In [73]:
# Print the values for your selected data for categorization

credit_score_data['family_status']

0                  married
1                  married
2                  married
3                  married
4        civil partnership
               ...        
17862              married
17863    civil partnership
17864              married
17865              married
17866              married
Name: family_status, Length: 17867, dtype: object

In [74]:
# Check the unique values
print(credit_score_data['family_status'].unique())

['married' 'civil partnership' 'divorced' 'unmarried' 'widow / widower']


Based on the unique values from the `family_status` column, I can identify two main groups: partnership and single. The partnership group include clients who are married or have civil partnership. The single group include clients who are divorced, unmarried, or widowed. Now let's categorize our data.

In [75]:
# Let's write a function to categorize the data based on common topics

def family_status_category(row):
    if row['family_status'] in ['married', 'civil partnership']:
        return 'partnership'
    else:
        return 'single'

In [76]:
# Create a column with the categories and count the values for them

# print(credit_score_data.loc[0])
# family_status_category(credit_score_data.loc[0])
credit_score_data['family_status'] = credit_score_data.apply(family_status_category, axis=1)

In [77]:
print(credit_score_data.head(10))
print(credit_score_data.tail(10))
# print(credit_score_data.groupby('family_status')['family_status'].count())

   children  days_employed  dob_years            education  education_id  \
0         1    8437.673028         42    bachelor's degree             0   
1         1    4024.803754         36  secondary education             1   
2         0    5623.422610         33  secondary education             1   
3         3    4124.747207         32  secondary education             1   
4         0     926.185831         27    bachelor's degree             0   
5         0    2879.202052         43    bachelor's degree             0   
6         0     152.779569         50  secondary education             1   
7         2    6929.865299         35    bachelor's degree             0   
8         0    2188.756445         41  secondary education             1   
9         2    4171.483647         36    bachelor's degree             0   

  family_status  family_status_id gender income_type  debt  total_income  \
0   partnership                 0      F    employee     0     40620.102   
1   partner

Next, let's look at the `children` column. 

In [78]:
# Looking through all the numerical data in your selected column for categorization
credit_score_data['children'].unique()

array([ 1,  0,  3,  2,  4, 20,  5])

In [79]:
# Getting summary statistics for the column

credit_score_data.groupby('children').size()

children
0     10869
1      4537
2      2022
3       322
4        40
5         9
20       68
dtype: int64

For the `children` column, I decide to create 3 categories based on these rules: 0 - none; 1,2 - low; 3,4,5 - medium; 20 - high. The reasons are simple, people without children should be a separate category. People with fewer children should be in one category, and people with more children should be in another category. Peole with manhy children should be in a distinct category. 

In [80]:
# Creating function for categorizing into different numerical groups based on ranges
def children_category(row):
    if row['children'] == 0:
        return 'none'
    elif row['children'] in [1,2]:
        return 'low'
    elif row['children'] in [3,4,5]:
        return 'medium'
    else:
        return 'high'

In [81]:
# Creating column with categories
credit_score_data['children'] = credit_score_data.apply(children_category, axis=1)

In [82]:
# Count each categories values to see the distribution
credit_score_data.groupby('children').size()

children
high         68
low        6559
medium      371
none      10869
dtype: int64

Next, let's look at the `total_income` column. 

In [83]:
# the five number summary of the total_income column

income_quartiles = np.percentile(credit_score_data['total_income'], [0,25, 50, 75, 100])
income_quartiles

array([  3418.824 ,  18133.5975,  23547.067 ,  32032.0105, 362496.645 ])

In [84]:
def income_category(row):
    if row['total_income'] < 18133.5975:
        return 'low'
    elif row['total_income'] >= 18133.5975 and row['total_income'] < 32032.0105:
        return 'medium'
    else:
        return 'high'

In [85]:
# Creating column with categories
credit_score_data['total_income'] = credit_score_data.apply(income_category, axis=1)

In [86]:
# Count each categories values to see the distribution
credit_score_data.groupby('total_income').size()

total_income
high      4467
low       4467
medium    8933
dtype: int64

Next, let's look at the `purpose` column. 

In [87]:
credit_score_data['purpose'].unique()

array(['purchase of the house', 'car purchase', 'supplementary education',
       'housing transactions', 'education', 'having a wedding',
       'purchase of the house for my family', 'buy real estate',
       'buy commercial real estate', 'buy residential real estate',
       'construction of own property', 'property', 'building a property',
       'buying my own car', 'buying a second-hand car',
       'to have a wedding', 'housing', 'transactions with my real estate',
       'cars', 'to become educated', 'getting an education',
       'second-hand car purchase', 'car', 'wedding ceremony',
       'to get a supplementary education', 'purchase of my own house',
       'real estate transactions', 'getting higher education',
       'transactions with commercial real estate', 'to own a car',
       'purchase of a car', 'profile education', 'university education',
       'to buy a car', 'buying property for renting out',
       'building a real estate', 'housing renovation',
       'going

In [88]:

real_estate = ['purchase of the house', 
             'housing transactions', 
             'purchase of the house for my family', 
             'buy real estate',
               'buy commercial real estate', 
               'buy residential real estate',
               'construction of own property', 
               'property', 
               'building a property', 'housing',
              'transactions with my real estate',
               'purchase of my own house',
               'real estate transactions',
               'transactions with commercial real estate',
               'buying property for renting out',
               'building a real estate', 'housing renovation'
              ]
car = ['car purchase', 'buying my own car', 
       'buying a second-hand car','cars',
      'second-hand car purchase', 'car', 
      'to own a car','purchase of a car',
       'to buy a car', 
      ]

education = ['supplementary education', 'education',
            'to become educated', 'getting an education',
             'to get a supplementary education',
             'getting higher education',
             'profile education', 'university education',
             'going to university'
            ]

wedding = ['having a wedding', 'to have a wedding',
          'wedding ceremony'
          ]

def purpose_category(row):
    if row['purpose'] in real_estate:
        return 'real_estate'
    elif row['purpose'] in car:
        return 'car'
    elif row['purpose'] in education:
        return 'education'
    elif row['purpose'] in wedding:
        return 'wedding'

In [89]:
credit_score_data['purpose'] = credit_score_data.apply(purpose_category, axis=1)

In [90]:
# Count each categories values to see the distribution
credit_score_data.groupby('purpose').size()

purpose
car            3559
education      3335
real_estate    9042
wedding        1931
dtype: int64

## Checking the Hypotheses


**Is there a correlation between having children and paying back on time?**

In [91]:
# Check the children data and paying back on time

debt_vs_children = credit_score_data.pivot_table(index='children', 
                              columns = 'debt',
                              values='dob_years', 
                              aggfunc= 'count', margins=True)
debt_vs_children
# Calculating default-rate based on the number of children

# debt_vs_children.loc[('high',0)]
# debt_vs_children.loc[('All','perct')] = debt_vs_children.loc[('All',1)]/debt_vs_children.loc[('All','All')]*100

for category in ['none', 'low', 'medium', 'high', 'All']:
    debt_vs_children.loc[(category,'perct_default')] = debt_vs_children.loc[(category,1)]/debt_vs_children.loc[(category,'All')]*100


debt_vs_children

debt          0     1    All  perct_default
children                                   
high         60     8     68      11.764706
low        5938   621   6559       9.467907
medium      341    30    371       8.086253
none       9977   892  10869       8.206827
All       16316  1551  17867       8.680808

**Conclusion**

For the entire dataset, the overall loan default rate is 8.68%. For people with no children and people with 3 to 5 children, their loan default rate is below average. But for people with 1 or 2 children, their loan default rate (9.47%) is a little above average. For people with lots of children, their loan default rate (11.76%) is significantly higher than average. Thus, there is a correlation between having children and paying back on time.    



**Is there a correlation between family status and paying back on time?**

In [92]:
# Check the family status data and paying back on time
debt_vs_family_status = credit_score_data.pivot_table(index='family_status', 
                              columns = 'debt',
                              values='dob_years', 
                              aggfunc= 'count', margins=True)


# Calculating default-rate based on family status

for category in ['partnership', 'single', 'All']:
    debt_vs_family_status.loc[(category,'perct_default')] = debt_vs_family_status.loc[(category,1)]/debt_vs_family_status.loc[(category,'All')]*100

debt_vs_family_status

debt               0     1    All  perct_default
family_status                                   
partnership    12744  1185  13929       8.507431
single          3572   366   3938       9.294058
All            16316  1551  17867       8.680808

**Conclusion**

For the entire dataset, the overall loan default rate is 8.68%. For people that are single, their loan default rate (9.29%) is a little above average. For people that have partners, their loan default rate (8.51%) is a little below average. Thus, there is a correlation between family status and paying back on time. 


**Is there a correlation between income level and paying back on time?**

In [93]:
# Check the income level data and paying back on time

debt_vs_income = credit_score_data.pivot_table(index='total_income', 
                              columns = 'debt',
                              values='dob_years', 
                              aggfunc= 'count', margins=True)

# Calculating default-rate based on income level

for category in ['low', 'medium', 'high', 'All']:
    debt_vs_income.loc[(category,'perct_default')] = debt_vs_income.loc[(category,1)]/debt_vs_income.loc[(category,'All')]*100


debt_vs_income

debt              0     1    All  perct_default
total_income                                   
high           4143   324   4467       7.253190
low            4058   409   4467       9.156033
medium         8115   818   8933       9.157058
All           16316  1551  17867       8.680808

**Conclusion**

For the entire dataset, the overall loan default rate is 8.68%. For people that have low or medium income, their loan default rate (9.16%) is a little above average. For people that have high income, their loan default rate (7.25%) is a significantly below average. Thus, there is a correlation between income level and paying back on time. 

**How does credit purpose affect the default rate?**

In [94]:
# Check the percentages for default rate for each credit purpose and analyze them

debt_vs_purpose = credit_score_data.pivot_table(index='purpose', 
                              columns = 'debt',
                              values='dob_years', 
                              aggfunc= 'count', margins=True)


for category in ['car', 'education', 'real_estate','wedding', 'All']:
    debt_vs_purpose.loc[(category,'perct_default')] = debt_vs_purpose.loc[(category,1)]/debt_vs_purpose.loc[(category,'All')]*100


debt_vs_purpose

debt             0     1    All  perct_default
purpose                                       
car           3202   357   3559      10.030908
education     3007   328   3335       9.835082
real_estate   8341   701   9042       7.752710
wedding       1766   165   1931       8.544795
All          16316  1551  17867       8.680808

**Conclusion**

For the entire dataset, the overall loan default rate is 8.68%. For people that borrow debt for wedding or real estate purposes, their loan default rate (8.54% and 7.75%) is a little below average. For people that borrow debt for education or car purposes, their loan default rate (9.84% and 10.03%) is higher than average. 
Thus, there is a correlation between credit purpose and default rate. 


# General Conclusion 

The analysis in this project discovered that the number of children a client has, marital status, total income level and purpose of debt all have association with debt default rate. 
